### ISSUE TITLE
[question] Validation and Datalad

### ISSUE THREAD

**Issue by vsoch** [datalad/datalad#2828]
This is another follow up question from:

https://github.com/datalad/datalad/issues/2814#issue-359426631

Does Datalad have any arm for validation, or is this up to the user? I am able to add a metadata template of sorts, is that checked against the data or just "I trust you" sort of deal? These seem to be the extractors. And they tell datalad how to parse files in a dataset, and I would guess they will skip over a file if it can't be parsed (but this is the extent of validation). What if I need multiple extractors within a single dataset, or want to define some other custom logic?

**Comment by yarikoptic-gitmate** [comment]
[GitMate.io](https://gitmate.io) thinks possibly related issues are https://github.com/datalad/datalad/issues/2824 ([question] DataLad Questions Set 1: Relationship to Git Annex), https://github.com/datalad/datalad/issues/1193 (datalad mv), https://github.com/datalad/datalad/issues/2693 (datalad aliases), https://github.com/datalad/datalad/issues/2825 ([question] levels of a dataset), and https://github.com/datalad/datalad/issues/227 (datalad get  barfes).

**Comment by mih** [comment]
I would be curious to learn about what kind of "custom logic" would be desirable, and how that would fit into our metadata model.

That being said: DataLad will not try to *understand* your metadata, it is more like a transport layer for metadata -- garbage in, garbage out  ;-)

**Comment by vsoch** [comment]
Ok cool - yes my understanding (and I agree with this!) is that validation is out of scope of datalad, and actually the bids validator is a perfect example of how validation should be. It does make sense that some custom validator might be integrated into datalad via a custom extractor OR just run on the dataset before interacting with datalad, period. BIDS validator is a good example for data in a browser (or on the filesystem) as is something like the [Google Schema Validator](https://developers.google.com/search/docs/guides/prototype) for the same metadata embedded in web pages. 

I think what we we would is to be able to:
 - read in a metadata description file (this is meta.json correct?)
 - parse over the root and then identify where data is missing, incorrectly formatted, or other.

Two questions: Do you think this could be provided generally for (a large number or many) schemas, and do you think the maximum amount of integration with datalad would be a plugin sort of deal? I think it would be ideal to be able to run it external to datalad or any other tool, e.g. let's say we have a folder at `root` with a bunch of subfolders, each with a "meta.json" file to describe content within:

```bash
./validate <root>
```
Should validate that. This setup would imply that:
 -  a single folder with a `meta.json` has the meta.json describing all content within
 -  for an exception we provided a custom `meta-<filename>-` file to describe `<file>`
 - or we have a filter within a specific `meta.json` to do the same

Thoughts?

**Comment by mih** [comment]
Nothing but one important thing would prevent a validation mechanism to be bolted on. That one thing is a machine-readable schema to base validation on. I am not talking about the technology, but real implementations for real standard. The BIDS validator would require its software stack to become a dependency (if that would be a feasible model). That quickly becomes infeasible when scaled up to 10-15 standards.

Note that `meta.json` is a outdated thing that @yarikoptic keeps around in the central dataset, please ignore it.

**Comment by vsoch** [comment]
yeah @yarikoptic told me! Thanks for the additional tidbits @mih. I've been quiet this weekend working on this --> https://github.com/BioSchemas/map2model/pull/35
in case you have comment! It's a template to produce a specification for bioschema, and the implementation I added **does** have validation, albeit for the specification files and not data they will describe. This would need to be done first before I create a dataset that would be managed by datalad and described by a bioschema.

**Comment by mih** [comment]
Closing "question"-type issue with a response, and no further activity.
